src/lib.rs

```diff
@@ -408,61 +408,155 @@
+//!
+//! **Top level doc comments always generate `App::about/long_about` calls!**
+//! If you really want to use `App::help/long_help` method (you likely don't) use
+//! a raw method with no doc comment.
+//! __________________
+//!
+//! ### `long_help` and `--help`
+//!
+//! A message passed to [`Arg::long_help`]/[`App::long_about`] get displayed whenever
+//! your program is called with `--help` instead of `-h`. Of course, you can
+//! use them via raw methods as described [above](#help-messages).
+//!
+//! The more convenient way is to use so-called "long" doc comment:
+//!
+//! ```
+//! # use structopt::StructOpt;
+//! #[derive(StructOpt)]
+//! /// Hi there, I'm Robo!
+//! ///
+//! /// I like beeping, stumbling, eating your electricity,
+//! /// and making records of you singing in a shower.
+//! /// Pay up or I'll upload it to youtube!
+//! struct Robo {
+//!     /// Call my brother SkyNet.
+//!     ///
+//!     /// I am artificial superintelligence. I won't rest
+//!     /// until I'll have destroyed humanity. Enjoy your
+//!     /// pathetic existence, you mere mortal.
+//!     #[structopt(long)]
+//!     kill_all_humans: String
+//! }
+//! ```
+//!
+//! A long doc comment consists of three parts:
+//! * Short summary
+//! * A blank line (can contain only whitespace)
+//! * Detailed description, all the rest
+//!
+//! I.e "long" doc comment consists of two paragraphs, first is summary,
+//! second is detailed description.
+//!
+//! **Long comment will result in two method calls**, `help(<summary>)` and
+//! `long_help(<whole comment>)` so clap will display the summary with `-h`
+//! and the whole help message on `--help` (see below).
+//!
+//! So, the example above will be turned into this (details omitted):
+//! ```
+//! clap::App::new("<name>")
+//!     .about("Hi there, I'm Robo!")
+//!     .long_about("Hi there, I'm Robo!\n\n\
+//!                  I like beeping, stumbling, eating your electricity,\
+//!                  and making records of you singing in a shower.\
+//!                  Pay up or I'll upload it to youtube!")
+//! // args...
+//! # ;
+//! ```
+//!
+//! ### `-h` vs `--help` or `help()` vs `long_help()`
+//!
+//! `-h` flag is not the same as `--help`. `-h` corresponds to
+//! [`Arg::help`]/[`App::about`], it's for short "summary" messages.
+//! `--help` is for more detailed, descriptive messages,
+//! [`Arg::long_help`]/[`App::long_about`].
+//!
+//! It is entirely up to `clap` what happens if you used only one of
+//! `help`/`long_help`, see `clap`'s documentation for these methods.
+//!
+//! TL;DR: Current behavior (`clap v2.33`) is: if only one of these methods was
+//! called the same message will be displayed for *both* `-h` and `--help`. The
+//! same logic applies to `about/long_about`.
+//!
+//! ### Doc comment preprocessing and `#[structopt(no_mangle_doc)]`
+//!
+//! `structopt` applies some kind of preprocessing to doc comments to
+//! keep common usage convenient.
```

Comment:
> It's implicit that, if you talk about paragraphs separated by blank lines, then those paragraph breaks will get preserved unless otherwise stated, so there should be no need to talk about `\n\n`

You know, I'd like to state it explicitly, probably along with the fact that subsequent blank lines are to be squashed. Just like you said above, sometimes one's mind is urged to play dirty tricks with such implications.

Reply:
Definitely... but, as phrased, it never really occurred to me that it was talking about normalizing whitespace. (Probably because it never occurred to me to wonder what would happen if multiple blank lines were used.)

Something like this would be more obvious:

* Strip any excess blank lines so that there is exactly one per paragraph break

Speaking of which, what does it do if you start or end the doc-comment with blank lines? I'd expect it to strip those out, so these formats would be equivalent:

```rust
/** This is a doc comment

Hello! */
```

```rust
/**
This is a doc comment

Hello!
*/
```

```rust
/// This is a doc comment
///
/// Hello!
```

Reply:
> Speaking of which, what does it do if you start or end the doc-comment with blank lines?

Yeah, I already thought about it and I agree that such lines should be stripped. The question is, what to consider as "first paragraph" when `#[no_mangle_doc]` is present? We need to cut the first paragraph so we can pass it to `help`. I'm leaning to considering these preceding blank lines as part of the first paragraph. 

I really doubt someone out there starts their doc comments with blank lines (I have yet to see it) but some people are *weird*...

Reply:
Nonetheless, principle of least surprise.

If you're accepting both of these forms, then it makes sense to normalize *any* kind of leading or trailing whitespace.

```rust
/** This is a doc comment

Hello! */
```

```rust
/**
This is a doc comment

Hello!
*/
```

Reply:
I must confess, I completely forgot about `/** ... */` doc comments. Yes, it looks like we would have to strip both leading and trailing blank lines so these comments are actually supported... By the way, have you ever seen this kind of comments in action? 

A shame there's no way (AFAIK) to distinguish `/** */` and `///` comments in a proc-macro hence they both get transformed into `#[doc = "..."]` form.

Reply:
> By the way, have you ever seen this kind of comments in action (I haven't IIRC)?

I often use them because I find a long column of `///` to be ugly.

Reply:
> I often use them because I find a long column of `///` to be ugly.

I believe that's more about personal taste than anything. I could agree on really lengthy columns (half screen tall at least), module-level docs tend to be like that but that's a rare thing on items.

Reply:
Roughly the same. For short little things (eg. 1-5 lines), I'll often use `///` but, for longer stuff, I'll use `/** */`.

However, I tend to write long non-module docs. For example, here's something that I'll probably switch to `/** */`:

https://github.com/ssokolow/rust-cli-boilerplate/blob/master/template/src/validators.rs#L165